File: Java solutions/Arrays/buySellStocks.java
class Solution {
    public int maxProfit(int[] prices) {
        

        //Initialize max profit to lowest possible value and minimum buy value to first element of prices array.
        // We do this because for max profit we always check for max between current profit and max profit whereas
        // for minimum buy price there has to be an initial buy price with which we compare upcoming prices and get min.

       int maxProfit = Integer.MIN_VALUE;
       int minBuy = prices[0];

       // Now we iterate over all values of prices array and set our minBuy and maxProfit at each iteration to the same value
       // or a new value based on min and max respectively.
       for(int i = 0; i < prices.length; i++)
       {
           int profit = prices[i] - minBuy;

           maxProfit = Math.max(profit,maxProfit);
           minBuy = Math.min(prices[i],minBuy);
       }

       // We finally return the maxProfit achieved after going through all values.
       return maxProfit;
    }
}

File: Python solutions/Arrays/buy_sell_stocks.py
from ast import List
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        # Initialize max_profit and min_buy
        max_profit = 0
        min_buy = prices[0]

        # Iterate over all nums in list and calculate profit at each step
        # considering sell price as current num and buy price as min_buy.
        # If profit is greater than max_profit update max_profit and if current price
        # is smaller than min_buy update min_buy
        
        for i in prices:
            profit = i - min_buy
            if profit >= max_profit:
                max_profit = profit
            
            if min_buy > i:
                min_buy = i
        
        return max_profit
